### ISSUE TITLE
PodAffinity predicate should be invalidated for podAdd

### ISSUE THREAD

**Issue by msau42** [kubernetes/kubernetes#67680]
**Is this a BUG REPORT or FEATURE REQUEST?**:
@kubernetes/sig-scheduling-bugs 

**What happened**:
In https://github.com/kubernetes/kubernetes/pull/67556, I am adding some new test cases with pod affinity and anti-affinity, and am seeing flaky test failures.

Debugging further, I see the following sequence of events from the test:
0. Pod A and Pod B have anti-affinity to each other using hostname topologyKey.
1. Pod A goes through the predicates successfully, fails during the volume assume (before pod assume)
2. Pod B goes through the predicates successfully, goes through pod assume and successfully binds to nodeA.
3. Pod A starts going through the scheduler again, it skips the PodAffinity predicate and uses the ecache value, which is true for all nodes.
4. Pod A gets bound to nodeA, violating the pod anti-affinity policy.

**What you expected to happen**:
PodAffinity predicate should be invalidated after Pod B is assumed.

**How to reproduce it (as minimally and precisely as possible)**:
From the PR, run repeatedly until failure:
```
make test-integration WHAT=./test/integration/scheduler GOFLAGS="-v" KUBE_TEST_ARGS="-race -run ^TestVolumeBindingWith"
```
It takes ~20 iterations to flake.  I added extra log statements in the PodAffinity predicate to determine that the predicate was not being called on the second run.

Edit: I updated the PR to disable the ecache for the test.  To reproduce, you'll have to modify the test to enable ecache again.

**Comment by resouer** [comment]
`InterPodAntiAffinity` invalidation is already there for Pod delete.

While, @misterikkit @bsalamat, I am a bit unsure if it's necessary to invalidate `InterPodAntiAffinity` for Pod add event. 

Here's the reason that I skipped its invalidation originally. Let's say:

1. pod_1 has AntiAffinity with any `app=nginx` Pod.
2. pod_1 is scheduled to Node X
3. pod_2 comes to scheduler after that, then we have two cases:
    1. pod_2 is equivalent to pod_1, so it will just use ecache and bound to Node X, ecache is still valid.
    2. pod_2 is not equivalent to pod1, then:
        1.  pod_2 is AntiAffinity against pod_1 (e.g. has` app=nginx` label), thus scheduler will reject it based on `InterPodAntiAffinity` rule, there will be no pod_2 add event for Node X. And no matter where pod_2 goes, ecache of that node is still valid.
       2. pod_2 is not AntiAffinity against pod_1 ( e.g. no` app=nginx` label), no matter where pod_2 goes, ecache of that node is still valid.

Thus, in my mind, as long as a non-equiv Pod which can break the cached AntiAffinity result will never be scheduled on that node, we don't need to do invalidation of it on pod add.

> Please feel free to correct me if I missed something here :D


As for the issue, it's actually caused by Pod A is considered as equivalent to itself during its re-scheduling. 

Of course invalidation `InterPodAntiAffinity` on Pod add can fix it, but then we will probably lose ecache speed up for `InterPodAntiAffinity` since we need to do invalidation for ALL nodes in the same failure domain for every Pod add.

I would prefer to bypass ecache for Pod re-scheduling , which is similar to [what we did in preemption phase](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/core/generic_scheduler.go#L555-L557). Because in my mind, re-scheduling normally means the Pod want to re-run predicates, and the scheduler is not clear what predicates it can reuse in ecache.

Please let me know if it's feasible and align with how equiv class is assumed to work.

**Comment by msau42** [comment]
I'm not sure I understand the 3.i. scenario where pod2 is equivalent to pod1.  Why is it ok for pod2 to be bound to the same node as pod1 if pod2 has anti-affinity on pod1?  I'm thinking about the case of a replicaset, where all the replicas use the same pod spec?

**Comment by k82cn** [comment]
>3.i pod_2 is equivalent to pod_1, so it will just use ecache and bound to Node X, ecache is still valid.

How about anti-affinity to itself?

> bypass ecache for Pod re-scheduling

seems reasonable

**Comment by resouer** [comment]
> How about anti-affinity to itself?

Emm ... time to eat my words. 

@msau42 I will send out a simple fix ASAP so you can move on with your PR. As we are migrating ecache to predicates pkg so I can ship a fine grained control there later.

**Comment by misterikkit** [comment]
I think this will be "fixed" by #71399

**Comment by misterikkit** [comment]
This can be closed as obsolete, as the old eCache implementation was removed. The new implementation will have a different architecture, but we will watch out for this class of bug.
/close
